### ISSUE TITLE
Snapcast with mopidy, audio not going through clients

### ISSUE THREAD

**Issue by gianfer14** [badaix/snapcast#784]
Hello! 

I am not sure whether this issue is more related to mopidy or snapcast, so please forgive me if I'm in the wrong place.

I am struggling for something which I hope is an error in my configuration, but at this point, I am not sure anymore. The audio is not going through the clients, but instead, all routed via the mopidy webpage (iris plugin) to the pc connected.

I am following the guidelines to set up multiroom audio, composed as follows:
1. server (docker with mopidy+snapcast), distributing audio to other snapcast clients (no sound will go out from it)
2. clients (raspberry4 + dac and speakers, with installed mopidy)
3. PC via snapcast web page or android device via app

During the setup, I managed to get this working configuration:
- Server streaming spotify music
- Raspberry client playing the server music
- PC connected on snapcast webpage playing the server music

Super, tried to use TTS from MPD (generating file somewhere else, sending to MPD and reproducing it) all went well and the voice came out as the music previously did.

So far all good. Then in the process of better understanding what happens if the clients are playing their own music while the snapcast server is running, I have lost the capability to stream to the clients. 

What happens now is the following (not what I was expecting):
- Server streaming music via spotify
- Raspberry client (snapcast server is set up and connected) plays music on the pc currently connected on mopidy web interface
- PC connected on snapcast webpage playing the server music (so gets 2 streams from 2 browser tabs)

What I did try already:
- [x] different browser, nothing changed
- [x] reset mopidy's settings, nothing changed
- [x] tried to send some noise from the server directly with cat /dev/urandom > /tmp/snapcast/snapfifo, noise worked as audio via pc, so nothing changed
- [x] reboot everything (server+clients+pc), as a last hope, nothing changed

Basically, my questions are:

1. how does the integration work between mopidy and snapcast? I couldn't find any documentation
2. how is it possible to get audio routed to the pc via mopidy's webpage? is it an issue with mopidy/iris?
3. does anyone has ever seen anything like this behavior? I couldn't find anything similar...
4. what should happen when a client playing music with mopidy, receives a streaming with snapcast begins to arrive?

Thanks for the patience and help you would give me, hope is all clear!

-gian

**Comment by badaix** [comment]
> how does the integration work between mopidy and snapcast? I couldn't find any documentation

https://github.com/badaix/snapcast/blob/master/doc/player_setup.md#mopidy

**Comment by mattmo92270** [comment]
HI @gianfer14 ,
What I understand from your post is that RPI clients should be able to play music from two sources : snapclient & local mopidy (not at the same time)
Have you looked into logs to find out if there is alsa issues like `alsa mixer/device busy`.
Your issue may come from your alsa configuration file in `/etc/asound.conf`.

**Comment by gianfer14** [comment]
Hi everyone, thank you for your replies!

@badaix, I have already used the information in the link you shared (https://github.com/badaix/snapcast/blob/master/doc/player_setup.md#mopidy), but I was mostly wondering how the priority is handled if a snapcast streaming is playing and a new mopify stream shows up, or the other way around. Is there any documentation regarding this?

@mattmo92270, correct, each RPI client should be able to play music from two sources (as a snapclient & via local mopidy, files etc). but for some reason after some correct test, the audio wasn't routed via the DAC connected, but via the webpage used to control it on my PC.

I will double check all the logs, and try to set a different debug option.
Where should I find issues with `alsa mixer/device busy`?

Thanks a lot!

**Comment by mattmo92270** [comment]
> Where should I find issues with alsa mixer/device busy?
Well I encountered once in snapclient logs using ‘--logfilter *:debug‘ option, an error stating something like "Alsa mixer busy" and sound of snapclient could not be heard through the DAC.
I managed to fix it by reconfiguring /etc/asound.conf

**Comment by gianfer14** [comment]
Hello, sorry for the late reply (amongst other things, my raspberry bootloader decided to die in the meantime, and I had to replace it).

I tried to look into the logs file, but I couldn't find anything related to the snapcast stream from mopidy/iris.

I tried to install and use snapcast client directly, setting up in `/etc/default/snapclient` the proper parameters, and it worked.

`/etc/asound.conf` is not present in my setup, I wonder if this culd be an issue even if the snapcast client is working as it should 
(in the mopidy setup for raspberry it suggeste a link `sudo ln -s /home/pi/.asoundrc /etc/asound.conf` but I didn't have none of the file existing)

I also tried to reinstall everithing, several times and with less and less possible conflicts, but still the behaviour is the same:
mopidy works (audio through raspberry's DAC) if I play music from file/spotify/youtube....etc, BUT if I click on the snapcast streaming, and try to enable it, the stream outputs from the PC/android I'm using....

I am really confused, also because in the first setup (which I cannot replicate unfortunately) it was working fine.

Now, am I still missing something? Or is this the way it should work?
If so,t it might be more an issue with Mopidy itself, what do you guys think?

Thanks a lot!!!!

**Comment by badaix** [comment]
> I tried to install and use snapcast client directly, setting up in /etc/default/snapclient the proper parameters, and it worked.

So this rather seems to be a setup issue with Iris/Mopidy.

**Comment by ChristianHin** [comment]
Sorry for reopening this issue. I think you are right badaix that it is related to Iris/Mopidy.
But @gianfer14 did you find a solution? it is complicated to describe this issue and you perfectly did it.

**Comment by gianfer14** [comment]
Hi @ChristianHin, I cannot remember if I found a workaround or a solution or is if the issue is still there... I can try to test it again in the next weeks and let you know, unfortunately I had to put it aside waiting to finish the house it will be placed in.

**Comment by skerit** [comment]
I've got the same problem still. Snapcast works fine, but Mopidy is not able to stream to it.

**Comment by kingosticks** [comment]
Your problem sounds like it would have been solved by the docs at https://github.com/badaix/snapcast/blob/master/doc/player_setup.md#mopidy in which case that isn't the OP's problem.

The original problem seemed to be someone trying to dynamically switch between playing audio from a locally running Mopidy server and snapclient. If I understand it right (not sure I do) then they are fundamentally looking for a way that two processes can use (write) to the same (audio) device at the same time. That's possible with pulse or with [ALSA dmix](https://raspberrypi.stackexchange.com/questions/89762/cannot-play-audio-from-two-processes-at-the-same-time). Or, they need to come up with a way for the two processes to co-operate and only one of them open the device at the same time. That's beyond the scope of both Mopidy and Snapcast.
